boskos/boskos.go

```diff
@@ -0,0 +1,473 @@
+			logrus.Warning("[BadRequest]method %v, expect GET", req.Method)
+			http.Error(res, "/metric only accepts GET", http.StatusMethodNotAllowed)
+			return
+		}
+
+		fmt.Fprint(res, "To be implemented.\n")
+	}
+}
+
+type Resource struct {
+	Type       string    `json:"type"`
+	Name       string    `json:"name"`
+	State      string    `json:"state"`
+	Owner      string    `json:"owner"`
+	LastUpdate time.Time `json:"lastupdate"`
+}
+
+type Ranch struct {
+	Resources   []Resource
+	lock        sync.RWMutex
+	storagePath string
+}
+
+func NewRanch(config string, storage string) (*Ranch, error) {
+
+	newRanch := &Ranch{
+		storagePath: storage,
+	}
+
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
+		if err == nil {
+			logrus.Infof("Current state: %v.", buf)
+			err = json.Unmarshal(buf, newRanch)
+			if err != nil {
+				return nil, err
+			}
+		} else if !os.IsNotExist(err) {
+			return nil, err
+		}
+	}
+
+	if err := newRanch.syncConfig(config); err != nil {
+		return nil, err
+	}
+
+	newRanch.logStatus()
+
+	return newRanch, nil
+}
+
+func (r *Ranch) logStatus() {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	for _, res := range r.Resources {
+		resJSON, _ := json.Marshal(res)
+		logrus.Infof("Current Resources : %v", string(resJSON))
+	}
+}
+
+func (r *Ranch) syncConfig(config string) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	file, err := ioutil.ReadFile(config)
+	if err != nil {
+		return err
+	}
+
+	var data []Resource
+	err = json.Unmarshal(file, &data)
+	if err != nil {
+		return err
+	}
+
+	return r.syncConfigHelper(data)
+}
+
+func (r *Ranch) syncConfigHelper(data []Resource) error {
+	// delete non-exist resource
+	valid := 0
+	for _, res := range r.Resources {
+		// If currently busy, yield deletion to later cycles.
+		if res.Owner != "" {
+			r.Resources[valid] = res
+			valid++
+			continue
+		}
+
+		for _, newRes := range data {
+			if res.Name == newRes.Name {
+				r.Resources[valid] = res
+				valid++
+				break
+			}
+		}
+	}
+	r.Resources = r.Resources[:valid]
+
+	// add new resource
+	for _, p := range data {
+		found := false
+		for _, exist := range r.Resources {
+			if p.Name == exist.Name {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			if p.State == "" {
+				p.State = "free"
+			}
+			r.Resources = append(r.Resources, p)
+		}
+	}
+	return nil
+}
+
+func (r *Ranch) saveState() {
+	if r.storagePath == "" {
+		return
+	}
+
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	// If fail to save data, fatal and restart the server
+	buf, err := json.Marshal(r)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error marshal ranch")
+	}
+	err = ioutil.WriteFile(r.storagePath+".tmp", buf, 0644)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error write file")
+	}
+	err = os.Rename(r.storagePath+".tmp", r.storagePath)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error rename file")
+	}
+}
+
+func (r *Ranch) Acquire(rtype string, state string, owner string) (int, *Resource) {
```

Comment:
This seems like an odd return signature. Specifically I do not think it makes sense to have the Ranch class understand http errors.

I would expect something more along the lines of: `Acquire()(*Resource, error)`. Ideally the errors are strongly typed.

Then I would expect the various `handle()` methods to map the strongly typed errors into status codes.